packages/api/lambdas/create-reconciliation-report.js

```diff
@@ -147,13 +150,95 @@ async function reconciliationReportForCollections() {
 }
 
 /**
- * Compare the granule holdings in CMR with Cumulus
+ * Compare the file holdings in CMR with Cumulus for a given granule
+ *
+ * @param {Object} granuleInDb - granule object in database
+ * @param {Object} granuleInCmr - granule object in CMR
+ * @param {Object} bucketsConfig - bucket configuration object
+ * @returns {Promise<Object>} an object with the okCount, onlyInCumulus, onlyInCmr
+ */
+async function reconciliationReportForGranuleFiles(granuleInDb, granuleInCmr, bucketsConfig) {
+  let okCount = 0;
+  const onlyInCumulus = [];
+  const onlyInCmr = [];
+
+  const granuleFiles = keyBy(granuleInDb.files, 'fileName');
+
+  // URL types for downloading granule files
+  const cmrGetDataTypes = ['GET DATA', 'GET RELATED VISUALIZATION'];
+  const cmrRelatedDataTypes = ['VIEW RELATED INFORMATION'];
+
+  // check each URL entry against database records
+  granuleInCmr.RelatedUrls.forEach((relatedUrl) => {
+    // only check URL types for downloading granule files and related data (such as documents)
+    if (cmrGetDataTypes.includes(relatedUrl.Type)
+      || cmrRelatedDataTypes.includes(relatedUrl.Type)) {
+      const urlFileName = relatedUrl.URL.split('/').pop();
+
+      // filename in both cumulus and CMR
+      if (granuleFiles[urlFileName] && bucketsConfig.key(granuleFiles[urlFileName].bucket)) {
+        // not all files should be in CMR
+        const accessUrl = constructOnlineAccessUrl({
+          file: granuleFiles[urlFileName],
+          distEndpoint: process.env.DISTRIBUTION_ENDPOINT,
+          buckets: bucketsConfig
+        });
+
+        if (accessUrl && relatedUrl.URL === accessUrl.URL) {
+          okCount += 1;
+        }
+        else if (cmrGetDataTypes.includes(relatedUrl.Type)) {
+          // ignore any URL which is not for getting data
+          // some files should not be in CMR such as private files
+          onlyInCmr.push({
+            URL: relatedUrl.URL,
+            Type: relatedUrl.Type,
+            GranuleUR: granuleInCmr.GranuleUR
+          });
+        }
+
+        delete granuleFiles[urlFileName];
+      }
+      else if (cmrGetDataTypes.includes(relatedUrl.Type)) {
+        // no matching database file, only in CMR
+        onlyInCmr.push({
+          URL: relatedUrl.URL,
+          Type: relatedUrl.Type,
+          GranuleUR: granuleInCmr.GranuleUR
+        });
+      }
+    }
+  });
+
+  // any remaining database items to the report
+  Object.keys(granuleFiles).forEach((fileName) => {
+    // private file only in database, it's ok
+    if (bucketsConfig.key(granuleFiles[fileName].bucket)
```

Comment:
So a private file that is not posted to CMR and only tracked by the Cumulus database is not considered "only in Cumulus" for the purpose of the reconciliation report?

Reply:
The private file is not supposed to be in CMR granule, and it's in Cumulus only, and shouldn't be reported as a discrepancy